lib/pharos/ssh/remote_file.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+require 'shellwords'
+require 'securerandom'
+
+module Pharos
+  module SSH
+    class RemoteFile
+      attr_reader :path
+      # Initializes an instance of a remote file
+      # @param [Pharos::SSH::Client]
+      # @param path [String]
+      def initialize(client, path)
+        @client = client
+        @path = path
+        freeze
+      end
+
+      alias to_s path
+
+      # Removes the remote file
+      def unlink
+        @client.exec!("rm #{escaped_path}")
+      end
+      alias rm unlink
+
+      def basename
+        File.basename(@path)
+      end
+
+      def dirname
+        File.dirname(@path)
+      end
+
+      def write(content)
+        tmp = temp_file_path.shellescape
+        @client.exec!(
+          "cat > #{tmp} && (sudo mv #{tmp} #{escaped_path} || (rm #{tmp}; exit 1))",
```

Comment:
This change looks reasonable. Have the error cases been tested?

Reply:
At least the most important error case seems to work as expected

```
$ echo 'foo' | ssh root@167.99.36.158 'cat > /tmp/foo && (sudo mv /tmp/foo /etc/foo || (rm /tmp/foo; exit 1))' && echo yay || echo nay
yay
$ echo 'foo' | ssh root@167.99.36.158 'cat > /tmp/foo && (sudo mv /tmp/foo /etc/foo/bar || (rm /tmp/foo; exit 1))' && echo yay || echo nay
mv: failed to access '/etc/foo/bar': Not a directory
nay
```

This does not leave any `/tmp/foo` behind.